### ISSUE TITLE
GraalVM's native-image substitute for `ClassLoader.getResource` normalizes resources' names while original doesn't

### ISSUE THREAD

**Issue by zakkak** [oracle/graal#4326]
**Describe the issue**
GraalVM's native-image substitute for `ClassLoader.getResource` normalizes resources' names while the original does not.

Running
```
        URL url = cl.getResource("META-INF/");
        URL url2 = cl.getResource("META-INF/./");
        System.out.println("url = " + url);
        System.out.println("url2 = " + url2);
```

in JVM results in:
```
url = jar:file:/home/zakkak/code/tmp/issue-reproducers/target/reproducer-1.0-SNAPSHOT.jar!/META-INF/
url2 = null
```

while in native-image it results in:
```
url = resource:META-INF
url2 = resource:META-INF
```

I would expect it to return:
```
url = resource:META-INF/
url2 = null
```
instead.

cc @jovanstevanovic @vjovanov as it relates to https://github.com/oracle/graal/pull/3597

**Steps to reproduce the issue**
```
cd /tmp
git clone --branch resources-with-trailing-slash https://github.com/zakkak/issue-reproducers reproducers
cd reproducers
mvn package
export JAVA_HOME=/opt/jvms/graalvm-ce-java11-22.0.0.2
$JAVA_HOME/bin/java \
  -jar target/reproducer-1.0-SNAPSHOT.jar
$JAVA_HOME/bin/native-image \
  -jar target/reproducer-1.0-SNAPSHOT.jar
./reproducer-1.0-SNAPSHOT
```

**Describe GraalVM and your environment:**
 - GraalVM version: **9b25754d49ca0fec016b1f4c03999f09fdb07404**
 - JDK major version: **11**
 - OS: **Fedora 35**
 - Architecture: **AMD64**

**Comment by adinn** [comment]
I'm not clear what solution you are expecting here. On the JVM the notion of a jar URL that identifies a resource in a jar on the classpath makes sense. In the context of a native image where there is no classpath, no jars and no loading of resources from a jar it doesn't really make any sense.

This is one of the areas where static and dynamic Java programs are necessarily going to have *variant* behaviour. Graal could go to great lengths to try to emulate JVM behaviour in an attempt to make programs work exactly the same however they are deployed. However, that may well turn out to be a dead end as indeed it will be in this case.

Imagine Graal hands you back a jar URL as per the JVM. What are you going to do with it other than use it to load the resource (i.e. beyond what you would do with the alternative URL Graal currently returns). If you try to decompose it, analyze its structure or use it to build/compare to other URLs  that is only going to get you into deeper trouble.

What is really needed here is not some pretend version of jar URLs that strives to retain continuity but rather a clear and coherent specification of what the various resource APIs do differently on Graal that the implementation can stick to and programmers can rely on.

**Comment by zakkak** [comment]
> I'm not clear what solution you are expecting here. On the JVM the notion of a jar URL that identifies a resource in a jar on the classpath makes sense. In the context of a native image where there is no classpath, no jars and no loading of resources from a jar it doesn't really make any sense.

You are right. What I expect is for native-image to retain the trailing slash in `url1` and fail to get a matching resource for `"META-INF/./"`. So IMO the expected output would be:

```
url = resource:META-INF/
url2 = null
```

**Comment by jovanstevanovic** [comment]
The thing is that Java makes a difference between accessing a resource when the app starts with `java -jar ...` and `java -cp ... <main class>`. In the first case, `ClassLoader.getResource` will make a canonical form out of the resource name. In the second case, it won't, as you show in this example.

In your case, your PR fixes the problem, but it very certainly generates other ones.
Adrian's final paragraph explains what we should do in situations like this.

**Comment by olpaw** [comment]
Apologies to @jovanstevanovic for prematurely merging https://github.com/oracle/graal/pull/4327.

This is actually more complicated than I anticipated.

It basically boils down to:

If a resource that should be embedded into the image comes from a classpath-entry that is

- a jar-file (`-cp myproj.jar`): then we should behave like @zakkak s PR.
- a directory: then we should behave like before (do the canonicalization).

**Comment by olpaw** [comment]
I will revert the PR and @jovanstevanovic will create a PR that will behave correctly in both cases (like described [above](https://github.com/oracle/graal/issues/4326#issuecomment-1055287779).

**Comment by zakkak** [comment]
@olpaw if https://github.com/oracle/graal/pull/4327 is not known to cause any issues, please let @jovanstevanovic revert the changes in his own PR, since at least it solves some issues for Quarkus.

**Comment by jovanstevanovic** [comment]
Actually, it will cause the issue right in the example that you provided (if you start it with java -cp ... <main class>). No worries, I'll try to make it work in both cases. Also, it would be valuable, if you have more examples that are not working, so I could pass through all of them once we solve the premier issue.

**Comment by zakkak** [comment]
> Actually, it will cause the issue right in the example that you provided (if you start it with `java -cp ... <main class>

Yes, I understood that part, but if there are no parties actually doing this with the nightly builds I am asking for you to let it be that way until the proper fix gets merged.

i.e. I am asking you to keep the `-cp myproj.jar` approach working instead of the `-cp myproj` approach, until we have a patch for making both work properly.

> No worries, I'll try to make it work in both cases. 

Thanks @jovanstevanovic 

> Also, it would be valuable, if you have more examples that are not working, so I could pass through all of them once we solve the premier issue.

I found out about this issue while investigating https://github.com/quarkusio/quarkus/issues/23604, so all the examples I am aware of are the ones included there and they are all using resources from jar files.

FWIW To reproduce it you will need to run:
```
git clone https://github.com/quarkusio/quarkus.git
./mvnw -Dquickly
export GRAALVM_HOME=path/to/your/graalvm/build
./mvnw -Dnative -pl integration-tests/vertx-web -Dtest-containers -Dstart-containers -Dnative.surefire.skip -Dformat.skip -Dno-descriptor-tests clean verify
./mvnw -Dnative -pl integration-tests/mailer -Dtest-containers -Dstart-containers -Dnative.surefire.skip -Dformat.skip -Dno-descriptor-tests clean verify
./mvnw -Dnative -pl integration-tests/resteasy-reactive-kotlin/standard -Dtest-containers -Dstart-containers -Dnative.surefire.skip -Dformat.skip -Dno-descriptor-tests clean verify
./mvnw -Dnative -pl integration-tests/amazon-lambda-http -Dtest-containers -Dstart-containers -Dnative.surefire.skip -Dformat.skip -Dno-descriptor-tests clean verify
./mvnw -Dnative -pl integration-tests/main -Dtest-containers -Dstart-containers -Dnative.surefire.skip -Dformat.skip -Dno-descriptor-tests clean verify
```

**Comment by zakkak** [comment]
Fixed by 7c4257f3f88d7cf2a368b92c714b8d55b44f36cd
